File: js/src/providers/keiro.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

const MODE: Record<string, string> = { fast: "light", balanced: "ai", deep: "deep" };

function firstString(...values: unknown[]): string | undefined {
  for (const value of values) {
    if (typeof value === "string" && value) return value;
  }
  return undefined;
}

function sourceItems(data: any): any[] {
  if (Array.isArray(data?.results)) return data.results;
  if (Array.isArray(data?.sources)) return data.sources;
  if (Array.isArray(data?.citations)) return data.citations;
  if (Array.isArray(data?.answer?.sources)) return data.answer.sources;
  if (Array.isArray(data?.data?.results)) return data.data.results;
  if (Array.isArray(data?.data?.sources)) return data.data.sources;
  return [];
}

function answerText(data: any): string | undefined {
  return firstString(data?.answer, data?.answer?.text, data?.answer?.content, data?.data?.answer);
}

function endpointFor(req: SearchRequest): string {
  if (req.includeContent) return "/api/v2/search/content";
  if (req.mode) return "/api/v2/search/fast";
  return "/api/v2/keirolabs";
}

export const keiro: ProviderSpec = {
  name: "keiro",
  envKeys: ["KEIRO_API_KEY"],
  baseUrlEnv: ["KEIRO_BASE_URL"],
  defaultBaseUrl: "https://kierolabs.space",
  capabilities: new Set(["mode", "content"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const endpoint = endpointFor(req);
    const body: Record<string, unknown> = {
      apiKey: ctx.apiKey,
      query: req.query,
      maxResults: req.maxResults,
    };
    if (endpoint !== "/api/v2/keirolabs") {
      body.mode = MODE[req.mode ?? ""] ?? "ai";
    }
    Object.assign(body, req.extra);
    return {
      method: "POST",
      url: `${ctx.baseUrl}${endpoint}`,
      headers: { "Content-Type": "application/json" },
      json: body,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const results = sourceItems(data).map((item: any) => {
      const snippet = firstString(item.snippet, item.description, item.content, item.text);
      return result({
        title: firstString(item.title, item.name),
        url: firstString(item.url, item.link, item.source_url),
        snippet,
        text: req.includeContent ? firstString(item.full_text, item.text, item.content, item.raw_content) : undefined,
        score: typeof item.score === "number" ? item.score : undefined,
        publishedDate: firstString(item.published_date, item.publishedDate, item.date),
        raw: item,
      });
    });
    return finalize(keiro, req, results, data, {
      answer: answerText(data),
      totalResults: data.total_results ?? data.totalResults,
      requestId: firstString(data.request_id, data.requestId, data.id),
      elapsedMs,
    });
  },
};

File: python/src/anysearch/providers/keiro.py
"""Keiro — source-cited web search via the v2 API."""

from __future__ import annotations

from typing import Any, Dict, Iterable, Optional

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability

_MODE = {"fast": "fast", "balanced": "balanced", "deep": "deep"}


def _first_string(*values: Any) -> Optional[str]:
    for value in values:
        if isinstance(value, str) and value:
            return value
    return None


def _source_items(data: Dict[str, Any]) -> Iterable[Dict[str, Any]]:
    if isinstance(data.get("results"), list):
        return data["results"]
    if isinstance(data.get("sources"), list):
        return data["sources"]
    if isinstance(data.get("citations"), list):
        return data["citations"]
    answer = data.get("answer")
    if isinstance(answer, dict) and isinstance(answer.get("sources"), list):
        return answer["sources"]
    nested = data.get("data")
    if isinstance(nested, dict):
        if isinstance(nested.get("results"), list):
            return nested["results"]
        if isinstance(nested.get("sources"), list):
            return nested["sources"]
    return []


def _answer_text(data: Dict[str, Any]) -> Optional[str]:
    answer = data.get("answer")
    if isinstance(answer, str):
        return answer
    if isinstance(answer, dict):
        return _first_string(answer.get("text"), answer.get("content"))
    nested = data.get("data")
    if isinstance(nested, dict):
        return _first_string(nested.get("answer"))
    return None


class KeiroProvider(BaseProvider):
    name = "keiro"
    env_keys = ("KEIRO_API_KEY",)
    base_url_env = ("KEIRO_BASE_URL",)
    default_base_url = "https://api.keiro.ai"
    capabilities = frozenset(
        {
            Capability.DOMAINS,
            Capability.DATE,
            Capability.MODE,
            Capability.ANSWER,
            Capability.CONTENT,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        body: Dict[str, Any] = {
            "query": req.query,
            "max_results": req.max_results,
            "mode": _MODE.get(req.mode or "", "balanced"),
            "source_citations": True,
        }
        if req.answer:
            body["answer"] = True
        if req.include_content:
            body["include_content"] = True
        if req.include_domains:
            body["include_domains"] = req.include_domains
        if req.exclude_domains:
            body["exclude_domains"] = req.exclude_domains
        if req.start_published_date:
            body["start_published_date"] = req.start_published_date[:10]
        if req.end_published_date:
            body["end_published_date"] = req.end_published_date[:10]
        body.update(req.extra)
        return PreparedRequest(
            method="POST",
            url=f"{self.base_url}/v2/source-cited-search",
            headers={"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"},
            json=body,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        results = []
        for item in _source_items(data):
            snippet = _first_string(
                item.get("snippet"),
                item.get("description"),
                item.get("content"),
                item.get("text"),
            )
            results.append(
                self._result(
                    title=_first_string(item.get("title"), item.get("name")),
                    url=_first_string(item.get("url"), item.get("link"), item.get("source_url")),
                    snippet=snippet,
                    text=_first_string(item.get("text"), item.get("content"), item.get("raw_content"))
                    if req.include_content
                    else None,
                    score=item.get("score") if isinstance(item.get("score"), (int, float)) else None,
                    published_date=_first_string(
                        item.get("published_date"), item.get("publishedDate"), item.get("date")
                    ),
                    raw=item,
                )
            )
        return self._response(
            req,
            results,
            data,
            answer=_answer_text(data),
            total_results=data.get("total_results") or data.get("totalResults"),
            request_id=_first_string(data.get("request_id"), data.get("requestId"), data.get("id")),
            elapsed_ms=elapsed_ms,
        )
